zkabacus-crypto/src/revlock.rs

```diff
@@ -87,42 +80,100 @@ pub struct RevocationLockCommitment(pub(crate) Commitment<G1Projective>);
 
 /// Commitment randomness corresponding to a [`RevocationLockCommitment`].
 #[derive(Debug, Clone, Copy, Serialize, Deserialize)]
-#[allow(missing_copy_implementations)]
 pub struct RevocationLockBlindingFactor(pub(crate) BlindingFactor);
 
-impl RevocationSecret {
-    /// Create a new, random revocation secret.
+/// An error in producing a revocation pair.
+#[derive(Debug, Copy, Clone, Error)]
+pub enum Error {
+    /// The provided revocation secret does not produce a valid revocation lock.
+    #[error("The revocation secret does not produce a valid revocation lock")]
+    InvalidSecret(),
+
+    /// The provided revocation secret does not produce the input revocation lock.
+    #[error("The revocation secret does not produce the provided revocation lock")]
+    MismatchedPair(),
+}
+
+impl TryFrom<UncheckedRevocationPair> for RevocationPair {
+    type Error = Error;
+
+    /// Try to convert an unchecked revocation pair into a verified revocation pair.
+    fn try_from(unchecked: UncheckedRevocationPair) -> Result<Self, Self::Error> {
+        let valid_pair = RevocationPair::try_from(unchecked.secret)?;
+
+        if unchecked.lock == valid_pair.lock {
+            Ok(valid_pair)
+        } else {
+            Err(Error::MismatchedPair())
+        }
+    }
+}
+
+impl TryFrom<UncheckedRevocationSecret> for RevocationPair {
+    type Error = Error;
+
+    /// Try to convert an unchecked revocation pair into a verified revocation pair.
+    fn try_from(unchecked: UncheckedRevocationSecret) -> Result<Self, Self::Error> {
+        // Compute the SHA3 hash of the byte representation of the unchecked revocation secret
+        // (this byte conversion must match `RevocationSecret::as_bytes()`)
+        let mut bytes = [unchecked.index; 33];
+        bytes[0..32].copy_from_slice(&unchecked.secret.to_bytes());
+        let digested = Sha3_256::digest(&bytes);
+
+        // Determine if the result is a valid revocation lock,
+        // ie is a Scalar in canonical form (smaller than the modulus)
+        let maybe_lock = Scalar::from_bytes(&<[u8; 32]>::try_from(&digested[..]).unwrap());
+
+        if maybe_lock.is_some().into() {
+            let valid_secret = RevocationSecret {
+                secret: unchecked.secret,
+                index: unchecked.index,
+            };
+            let valid_lock = RevocationLock(maybe_lock.unwrap());
+            Ok(RevocationPair {
+                secret: valid_secret,
+                lock: valid_lock,
+            })
+        } else {
+            Err(Error::InvalidSecret())
+        }
```

Comment:
What do you think of:

```rust
match maybe_scalar.into() {
    Some(scalar) => Ok(RevocationPair {
        secret: RevocationSecret {
            secret: unchecked.secret,
            index: unchecked.index,
        },

        lock: RevocationLock(scalar),
    }),

    None => Err(Error::InvalidSecret),
}
```
?

Also, is it important that this particular conversion is done in constant time? I think when we use `into()` we lose this guarantee. See https://docs.rs/subtle/2.4.1/subtle/struct.CtOption.html#method.from

Reply:
I don't think it matters if this is constant time. @marsella?

Reply:
I think this is the point where the behavior changes depending on the value of the `CtOption`, so we have to break out of the constant time types.

Reply:
Concretely, what are you suggesting here?

Reply:
I think we should use Mukund's suggestion exactly as described above. This leaves the `into()`, which breaks the constant-time behavior of the `maybe_lock`. We don't need constant time guarantees here.